src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -499,8 +499,7 @@ ObjectDataHandler::write_ret do_removals(
       return ctx.tm.remove(
 	ctx.t,
 	pin->get_key()
-      ).si_then(
-	[](auto){},
+      ).discard_result().handle_error_interruptible(
```

Comment:
how about adding an assertion that returned refcount is always zero?

Reply:
Seems the returned refcount can be non-zero if there are other valid clones?

Reply:
The returned refcount was from the indirect mapping, whose refcount should always be 1 before removing. The refcount of intermediate mapping doesn't return to transaction manager.

https://github.com/ceph/ceph/blob/a2e5d82920b689d73c3fb643592dc410d7e6b0df/src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc#L703-L718

both L707 and L714 are from indiect mapping itself.

Reply:
the unit test shows that this assertion will break the invariant of cloning process:
https://github.com/ceph/ceph/blob/a2e5d82920b689d73c3fb643592dc410d7e6b0df/src/crimson/os/seastore/object_data_handler.cc#L1740-L1744

Reply:
> The returned refcount was from the indirect mapping, whose refcount should always be 1 before removing. The refcount of intermediate mapping doesn't return to transaction manager.

@zhscn I think returning the refcount from indirect mapping isn't meaningful/useful. It is better to return the actual refcount.

Reply:
> the unit test shows that this assertion will break the invariant of cloning process:

Yeah, it's decreacing the original ref-counters which has become intermediate, so the ref-counter isn't 0.